backend/app/adapter/sio_srv.py

```diff
@@ -0,0 +1,73 @@
+import logging
+import asyncio
+import os
+
+from typing import Dict
+import socketio
+from hypercorn.config import Config
+from hypercorn.asyncio import serve
+
+toplvl_namespace = "/game"
+srv = socketio.AsyncServer(async_mode="asgi")
+_logger = logging.getLogger(__name__)
+
+
+@srv.on("chat", namespace=toplvl_namespace)
```

Comment:
考慮使用`config.py `去管理常量
檔案內可能包含了
- 伺服器主機/連接埠
- namespace/event name
- 日誌記錄配置
- 任何其他常數

如果是參考我這邊說的[建議](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/issues/37#issuecomment-1825934152)
> 明確分隔傳送的是遊戲狀態更新(`game`)還是聊天訊息(`chat`)

其中namespace/event name 可能長這樣
```python
# config.py
NAMESPACE = "/game" 

CHAT_EVENT= "chat"
GAME_RESPONSE_EVENT = "game"

INIT_EVENT = "init"

DEINIT_EVENT = "deinit"
``` 

我注意到實際上程式碼在我上述例子中應該為
```python
# config.py

...

INIT_EVENT = "init"
INIT_RESPONSE_EVENT = "init"

DEINIT_EVENT = "deinit"
DEINIT_RESPONSE_EVENT = "deinit"
``` 
我想這應該是考慮就算為同一遊戲房間的玩家，
`chat`頻道為強制接收，
其他可依照偏好選擇是否接收？

Reply:
> 是否考慮使用socketio_config.py 去管理常量

It is a good idea , I was thinking of this after finishing the [TODO list](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/issues/37#issuecomment-1842635038), but it is also OK if I get this done now.

 It would be good to split your example to :
- `constant.py` , focus on constants, anything which shouldn't be modified by user (more precisely, server admin) , such as `socket.io` namespace(s) , all event names , maybe `GameErrror`
- `config.py` , anything configurable will be placed at here, such as host domain / port of REST server and `socket.io` server , logging file path ...etc

 
> 我想這應該是考慮就算為同一遊戲房間的玩家，
`chat`頻道為強制接收，其他可依照偏好選擇是否接收？

All the 3 events : `chat`, `init` , `deinit`, should be visible only to all players of the same room (same game ID) .
not sure what you mean by the word -> 其他可依照偏好選擇是否接收？

Reply:
> `chat`頻道為強制接收，其他可依照偏好選擇是否接收？

讓我盡可能詳細說明我原先的疑問。
我注意到程式碼中,`chat `事件似乎是廣播給同一遊戲房間中的所有玩家。
但是對於其他像 `init`和 `deinit`事件中,
其
`@srv.emit(RESPONSE_EVENT)`
相對應的`RESPONSE_EVENT`是不同的，並非統一的使用`game`
我理解為這是因為隱含了一個設計意圖
目的是讓玩家可以選擇性地決定是否接收這些事件
想確認是否理解正確

舉例來說:
- 玩家 A 和 玩家 B 在同一個遊戲房間，玩家 Z 在另一個不同的遊戲房間
  - 玩家 A 和 玩家 B 會同時接收相同的 `chat`事件，玩家 Z 則不會
  - 但是玩家 A,B 可以選擇接收  `init`/`deinit` 事件
    -  例如玩家 A 選擇性的僅接收  `init` 而玩家 B 僅接收`deinit`
    -  或者玩家 A 選擇接收 `init`/`deinit` 玩家 B選擇皆不接收 

如果事件的處理邏輯很相似,
通常考慮使用相同的回應事件(本例中為`game`),
這樣可以簡化前端的處理邏輯
但是這邊考量
> 明確分隔傳送的是遊戲狀態更新(game)還是聊天訊息(chat)

所以分成兩個`game`與`chat`

若是每個事件需要返回不同資料給前端
例如我猜想這邊設計意圖是讓玩家可自由選擇是否接收某些事件或者接收詳細程度
那麼應該就應使用不同的回應事件(本例中為 `init`,`deinit` ),以套用不同的事件類型
那這邊就會分成`chat`與 `init`、`deinit`、.... 等

以上使用那個方案,沒有絕對標準答案,需要根據構想的情況做權衡和設計

Reply:
> 我注意到程式碼中,`chat `事件似乎是廣播給同一遊戲房間中的所有玩家。
但是對於其他像 `init`和 `deinit`事件中,
其 `@srv.emit(RESPONSE_EVENT)`
相對應的`RESPONSE_EVENT`是不同的，並非統一的使用`game`

- 目前查看 branch `feature/backend/socketio/chat` 中的 [server code ](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/blob/feature/backend/socketio/chat/backend/app/adapter/sio_srv.py) , 我用 `srv.on(...)` decorator  註冊 3 個events `chat`, `init`, `deinit` 都在同一個 namespace `/game` 底下, 所以都有統一使用 `/game`.
  - 或者您能否指出程式碼的哪一部分, 讓您認為它沒有綁定到 namespace `/game` ?
- 補充說明,  為了方便理解, server side 發出的 event name 應取決於 (等於) client side 發出的 event name, 例如
  - client side 發出事件 `init` , server side 收到後也應該發出同名 `init` 的事件

> 我理解為這是因為隱含了一個設計意圖
> 目的是讓玩家可以**選擇性地決定**是否接收這些事件

其實我沒有計劃這麼做.

目前我的做法是: 讓同一個房間的所有玩家都收到所有事件
考慮這遊戲 (pandemic reign-of-Cthulhu) 的應用場景, 我們應該不需要實作這個細節

(c.c. @wingtkw)

> 舉例來說:
>  - 玩家 A 和 玩家 B 在同一個遊戲房間，玩家 Z 在另一個不同的遊戲房間
>    - 玩家 A 和 玩家 B 會同時接收相同的 chat事件，玩家 Z 則不會

補充, 玩家 Z 不會收到來自其他房間的**任何**事件 , 不只是 `chat`

>    -  但是玩家 A,B 可以選擇接收 init/deinit 事件
>         - 例如玩家 A 選擇性的僅接收 init 而玩家 B 僅接收deinit
>         - 或者玩家 A 選擇接收 init/deinit 玩家 B選擇皆不接收

所以, 玩家 A 發出事件 `init`, 玩家 B 也將收到 事件 `init` ,
目前我沒有考慮讓 玩家 B 自由控制是否接收某些事件 e.g.  `init` and `deinit`